### PR TITLE
🎨 add no delegatecall modifier in mainreg

### PR DESCRIPTION
**Pull request by j-vp**

### CODE REVIEW

File: src/AssetRegistry/MainRegistry.sol [hunk 1]
```diff
@@ -188,7 +196,7 @@ contract MainRegistry is Ownable {
         address[] calldata assetAddresses,
         uint256[] calldata assetIds,
         uint256[] calldata amounts
-    ) public onlyVault {
+    ) public onlyVault noDelegate {
```
Comment (Thomas-Smets) [inline src/AssetRegistry/MainRegistry.sol:199]: Add test plz

File: src/AssetRegistry/MainRegistry.sol [hunk 2]
```diff
@@ -211,7 +219,11 @@ contract MainRegistry is Ownable {
      * @param amounts An array of amounts to be withdrawn
      * @dev batchProcessWithdrawal in the pricing module updates the maxExposure
      */
-    function batchProcessWithdrawal(address[] calldata assetAddresses, uint256[] calldata amounts) public onlyVault {
+    function batchProcessWithdrawal(address[] calldata assetAddresses, uint256[] calldata amounts)
+        public
+        onlyVault
+        noDelegate
```
Comment (Thomas-Smets) [inline src/AssetRegistry/MainRegistry.sol:225]: Add test plz

File: src/test/6_MainRegistry.t.sol [hunk 3]
```diff
@@ -1038,3 +1038,142 @@ contract PricingLogicTest is MainRegistryTest {
...
+        vm.startPrank(proxyAddr);
+        vm.expectRevert("Delegate calls not allowed.");
+        (bool success,) = address(mainRegistry).delegatecall(
+            abi.encodeWithSignature(
+                "batchProcessWithdrawal(address[] calldata,uint256[] calldata)", assetAddresses, assetAmounts
+            )
+        );
+        vm.stopPrank();
+
+        success; //avoid warning
```
Comment (Thomas-Smets) [inline src/test/6_MainRegistry.t.sol:1177]:
```suggestion
```

File: src/test/6_MainRegistry.t.sol [hunk 4]
```diff
@@ -1038,3 +1038,142 @@ contract PricingLogicTest is MainRegistryTest {
...
+
+        uint256[] memory assetIds = new uint256[](1);
+        assetIds[0] = 0;
+
+        uint256[] memory assetAmounts = new uint256[](1);
+        assetAmounts[0] = amountLink;
+
+        vm.startPrank(proxyAddr);
+        vm.expectRevert("Delegate calls not allowed.");
+        (bool success,) = address(mainRegistry).delegatecall(
```
Comment (Thomas-Smets) [inline src/test/6_MainRegistry.t.sol:1170]:
```suggestion
        address(mainRegistry).delegatecall(
```

File: src/test/6_MainRegistry.t.sol [hunk 5]
```diff
@@ -1038,3 +1038,142 @@ contract PricingLogicTest is MainRegistryTest {
...
+            abi.encodeWithSignature(
+                "batchProcessDeposit(address[] calldata,uint256[] calldata,uint256[] calldata)",
+                assetAddresses,
+                assetIds,
+                assetAmounts
+            )
+        );
+        vm.stopPrank();
+
+        success; //avoid warning
```
Comment (Thomas-Smets) [inline src/test/6_MainRegistry.t.sol:1126]:
```suggestion
```

File: src/test/6_MainRegistry.t.sol [hunk 6]
```diff
@@ -1038,3 +1038,142 @@ contract PricingLogicTest is MainRegistryTest {
...
+
+        uint256[] memory assetIds = new uint256[](1);
+        assetIds[0] = 0;
+
+        uint256[] memory assetAmounts = new uint256[](1);
+        assetAmounts[0] = amountLink;
+
+        vm.startPrank(proxyAddr);
+        vm.expectRevert("Delegate calls not allowed.");
+        (bool success,) = address(mainRegistry).delegatecall(
```
Comment (Thomas-Smets) [inline src/test/6_MainRegistry.t.sol:1116]:
```suggestion
        address(mainRegistry).delegatecall(
```

File: src/test/6_MainRegistry.t.sol [hunk 7]
```diff
@@ -1038,3 +1038,142 @@ contract PricingLogicTest is MainRegistryTest {
...
+                baseCurrencyLabel: "ETH",
+                baseCurrencyUnitCorrection: uint64(10 ** (18 - Constants.ethDecimals))
+            })
+        );
+        mainRegistry.addPricingModule(address(standardERC20PricingModule));
+        mainRegistry.addPricingModule(address(floorERC721PricingModule));
+        standardERC20PricingModule.addAsset(address(eth), oracleEthToUsdArr, emptyRiskVarInput, type(uint128).max);
+        standardERC20PricingModule.addAsset(address(link), oracleLinkToUsdArr, emptyRiskVarInput, type(uint128).max);
+
+        floorERC721PricingModule.addAsset(
```
Comment (Thomas-Smets) [inline src/test/6_MainRegistry.t.sol:1070]: Why add bayc?

File: src/test/6_MainRegistry.t.sol [hunk 8]
```diff
@@ -1038,3 +1038,142 @@ contract PricingLogicTest is MainRegistryTest {
...
+        address[] memory assetAddresses = new address[](1);
+        assetAddresses[0] = address(link);
+
+        uint256[] memory assetIds = new uint256[](1);
+        assetIds[0] = 0;
+
+        uint256[] memory assetAmounts = new uint256[](1);
+        assetAmounts[0] = amountLink;
+
+        (, uint128 oldExposure) = standardERC20PricingModule.exposure(address(link));
```
Comment (Thomas-Smets) [inline src/test/6_MainRegistry.t.sol:1093]: No assets deposited yet, why store oldExposure?

File: src/test/6_MainRegistry.t.sol (hunk 3 above)
Comment (j-vp) [reply src/test/6_MainRegistry.t.sol:1177]: will always have a 'unused variable' warning

File: src/test/6_MainRegistry.t.sol (hunk 4 above)
Comment (j-vp) [reply src/test/6_MainRegistry.t.sol:1170]: will throw a 'return of delegatecall not evaluate' or something similar

File: src/test/6_MainRegistry.t.sol [hunk 9]
```diff
@@ -1038,3 +1038,142 @@ contract PricingLogicTest is MainRegistryTest {
         assertEq(expectedLiquidationThreshold, actualLiquidationThreshold);
     }
 }
+
+contract DelegateTest is MainRegistryTest {
```
Comment (Thomas-Smets) [inline src/test/6_MainRegistry.t.sol:1042]: Add tests under ASSET MANAGEMENT in the testfile please, we are testing batchProcessWithdrawal and batchProcessDeposit

File: src/test/6_MainRegistry.t.sol (hunk 5 above)
Comment (j-vp) [reply src/test/6_MainRegistry.t.sol:1126]: will throw a 'return of delegatecall not evaluate' or something similar

File: src/test/6_MainRegistry.t.sol (hunk 6 above)
Comment (j-vp) [reply src/test/6_MainRegistry.t.sol:1116]: will throw a 'return of delegatecall not evaluate' or something similar

File: src/test/6_MainRegistry.t.sol (hunk 8 above)
Comment (j-vp) [reply src/test/6_MainRegistry.t.sol:1093]: more generic than hardcoding 0

File: src/test/6_MainRegistry.t.sol (hunk 9 above)
Comment (j-vp) [reply src/test/6_MainRegistry.t.sol:1042]: we are actually testing an authorization modifier that by chance is on the batchprocess functions

Comment (Thomas-Smets) [reply src/test/6_MainRegistry.t.sol:1042]: No we are testing the functions batchProcessWithdrawal and batchProcessDeposit that have a modifier, we do it the same way for all other modifiers such as onlyOwner, onlyVault etc.

File: src/test/6_MainRegistry.t.sol [hunk 10]
```diff
@@ -1038,3 +1038,137 @@ contract PricingLogicTest is MainRegistryTest {
...
+        mainRegistry.addPricingModule(address(standardERC20PricingModule));
+        standardERC20PricingModule.addAsset(address(eth), oracleEthToUsdArr, emptyRiskVarInput, type(uint128).max);
+        standardERC20PricingModule.addAsset(address(link), oracleLinkToUsdArr, emptyRiskVarInput, type(uint128).max);
+
+        mainRegistry.setFactory(address(factory));
+
+        vm.stopPrank();
+
+        vm.startPrank(vaultOwner);
+        proxyAddr = factory.createVault(uint256(keccak256(abi.encodePacked(block.timestamp, msg.sender))), 1);
```
Comment (Thomas-Smets) [inline src/test/6_MainRegistry.t.sol:1074]:
```suggestion
        proxyAddr = factory.createVault(uint256(keccak256(abi.encodePacked(block.timestamp, msg.sender))), 0);
```

**Review by Thomas-Smets — APPROVED** [review]
gtg

**Pull request merged**